src/include/schema.h

```diff
@@ -79,27 +79,37 @@ struct __wt_table {
 #define	WT_COLGROUPS(t)	WT_MAX((t)->ncolgroups, 1)
 
 /*
+ * WT_WITH_LOCK_WAIT --
+ *	Wait for a lock, perform an operation, drop the lock.
+ */
+#define	WT_WITH_LOCK_WAIT(session, lock, flag, op) do {			\
+	if (F_ISSET(session, (flag))) {					\
+		op;							\
+	}  else {							\
+		__wt_spin_lock(session, (lock));			\
+		F_SET(session, (flag));					\
+		op;							\
+		F_CLR(session, (flag));					\
+		__wt_spin_unlock(session, (lock));			\
+	}								\
+} while (0)
+
+/*
  * WT_WITH_LOCK --
  *	Acquire a lock, perform an operation, drop the lock.
  */
 #define	WT_WITH_LOCK(session, ret, lock, flag, op) do {			\
 	ret = 0;							\
-	if (F_ISSET(session, (flag))) {					\
-		op;							\
-	} else if (F_ISSET(session, WT_SESSION_LOCK_NO_WAIT)) {		\
+	if (!F_ISSET(session, (flag)) &&				\
+	    F_ISSET(session, WT_SESSION_LOCK_NO_WAIT)) {		\
```

Comment:
FTR, this is structured this way to allow for reuse of the `WT_WITH_LOCK_WAIT` macro, rather than repeating that logic.